Make SPI return buffers upon error

alexandruradovici opened the pull request:
### Pull Request Overview

This pull request changes the SPI HIL so that function return the buffers to the caller in case of errors. It also adds deferred callbacks to the SPI Mux, similar to what the I2C Mux.

I have two questions that need some feedback:
1. Several drivers now output a warning due to unverified result. The two simple solutions are: use `let _ =` or use `.unwrap()`. I'm not sure which one would be best.
2. The setup functions, like `configure` or `set_rate` do have to return results,  as the Mux might fail to execute them. There is no callback function defined for this. Should I add a new callback function, something like `command_done` or simply remove the return from the functions?

### Testing Strategy

The PR still needs testing and review.

### TODO or Help Wanted

Testing on different boards would be very helpful. I can test on STM32F4 and Adafruit CLUE.

### Documentation Updated

- [ ] Updated the relevant files in `/docs`, or no updates are required.

### Formatting

- [ ] Ran `make prepush`.

alistair23 commented:
The HIL changes all look good to me.

1. I don't think `let _ = ` is a great option, although `unwrap()` also isn't good. We should try and gracefully handle the error.
2. Things like `configure` and `set_rate` should return an error (what you have here is correct) and I would say they don't need a callback as they are going to be synchronous operations,

alexandruradovici commented:
> The HIL changes all look good to me.
> 
>     1. I don't think `let _ = ` is a great option, although `unwrap()` also isn't good. We should try and gracefully handle the error.
> 

None of them are fine from this point of view, but changing all drivers is a long process, I was looking for a quick solution. Technically none of these tow would change any behavior, as drivers do not check for errors now. I could try to fix most of them, but some drivers need to be heavily modified to report the errors.
>     2. Things like `configure` and `set_rate` should return an error (what you have here is correct) and I would say they don't need a callback as they are going to be synchronous operations,

The problem is when you use them through a mux, the mux queries the op vut does not execute it immediately.

alistair23 commented:
I guess `unwrap()` might be better then as it's hopefully easier to remove in the future (compared to a `let _ = ` which can just silently fail).

Shouldn't the mux call `configure()` then start a transaction, so it ends up being called immediately?

alexandruradovici commented:
> I guess `unwrap()` might be better then as it's hopefully easier to remove in the future (compared to a `let _ = ` which can just silently fail).
> 
> Shouldn't the mux call `configure()` then start a transaction, so it ends up being called immediately?

The VirtualSpiMasterDevice schedules an `Op::Configure` and the mux will execute it when it can. This might be immediately or not. The virtual device's `configure` can only check if it has another outstanding Op, otherwise any other error will be checked when the mux actually execute the configure.

hudson-ayers commented:
> Several drivers now output a warning due to unverified result. The two simple solutions are: use let _ = or use .unwrap(). I'm not sure which one would be best.

I get Alistair's argument that `unwrap()` at least does not fail silently, but in the large scale changeover to using `Result` we generally used the `let = ` approach. Adding `unwrap()` can lead to lots of kernel panics in places where errors occurred but weren't causing us many problems. Without auditing each of those call sites it can be hard to know whether the error was ignored because there was no easy way to report it or because errors are expected to be impossible.

alexandruradovici commented:
> > Several drivers now output a warning due to unverified result. The two simple solutions are: use let _ = or use .unwrap(). I'm not sure which one would be best.
> 
> I get Alistair's argument that `unwrap()` at least does not fail silently, but in the large scale changeover to using `Result` we generally used the `let = ` approach. Adding `unwrap()` can lead to lots of kernel panics in places where errors occurred but weren't causing us many problems. Without auditing each of those call sites it can be hard to know whether the error was ignored because there was no easy way to report it or because errors are expected to be impossible.

I'll try to go driver by driver and add error checking. For drivers like `sdcard` where this is more complex, I'll just silent fail it with `let _ =`.

lschuermann commented:
> I'll try to go driver by driver and add error checking. For drivers like `sdcard` where this is more complex, I'll just silent fail it with `let _ =`.

That'd be awesome! If it's not too much of a hassle for you, maybe you can split adding error handling up into different commits by driver? Adding proper error handling is something we definitely want, and having separate commits makes it significantly easier to review & ensure that there are no fundamental changes in the logic.

alexandruradovici commented:
> > I'll try to go driver by driver and add error checking. For drivers like `sdcard` where this is more complex, I'll just silent fail it with `let _ =`.
> 
> That'd be awesome! If it's not too much of a hassle for you, maybe you can split adding error handling up into different commits by driver? Adding proper error handling is something we definitely want, and having separate commits makes it significantly easier to review & ensure that there are no fundamental changes in the logic.

I will split them up in separate commits. I think I do need this to be merged first. Should I put unwrap and a command with TODO above each line?

lschuermann commented:
I suppose using `let _ =` is the favored approach if you want this PR to be merged standalone. Relevant PR which required quite similar changes: https://github.com/tock/tock/pull/2508

phil-levis reviewed (commented):
As an API change, this looks good to me.

alexandruradovici commented:
> I suppose using `let _ =` is the favored approach if you want this PR to be merged standalone. Relevant PR which required quite similar changes: #2508

This PR was my inspiration for the suggestion :)

alexandruradovici commented:
> > I suppose using `let _ =` is the favored approach if you want this PR to be merged standalone. Relevant PR which required quite similar changes: #2508
> 
> This PR was my inspiration for the suggestion :)

alexandruradovici commented:
I'll make the changes to add `let _ = ` and mark the lines with TODO and get started with the drivers.

hudson-ayers commented:
> I'll make the changes to add `let _ = ` and mark the lines with TODO and get started with the drivers.

Thanks!

alexandruradovici commented:
I added error returns for all SPI HIL functions. Here are a few questions that are still open.

1. What happens if the `init` function fails? This is used mostly in components. 
2. I modified the `sam4l`'s `specify_chip_select` function to report an error if the chip select peripheral is wrong, @phil-levis can you take a look at it as you are the author of the file?
3. The SPI Mux might not be correct for `configure`-like functions. Each time a virtual device asks a configure option, the Mux queues the action and executes it as soon as the real device is idle. Shouldn't these configurations be issues before each data transfer? As it is now, each virtual device can specify a different configuration, but the latest one will be used by all.

alistair23 commented:
>     1. What happens if the `init` function fails? This is used mostly in components.

You might not want to return a `Result` from `init()` as generally the init function can't fail

alexandruradovici commented:
> > ```
> > 1. What happens if the `init` function fails? This is used mostly in components.
> > ```
> 
> You might not want to return a `Result` from `init()` as generally the init function can't fail

On some boards `init` performs `configure` which can fail. Should we just ignore that?

alistair23 commented:
Hmm... We shouldn't ignore errors. Could we move the configure call to happen when another operation is started? Then we can report the error.

Otherwise I guess a `panic`/`unwrap`, as we have no one to report the error to.

phil-levis commented on kernel/src/hil/spi.rs:
```diff
@@ -91,22 +92,22 @@ pub trait SpiMaster {
     fn get_rate(&self) -> u32;
-    fn set_clock(&self, polarity: ClockPolarity);
+    fn set_clock(&self, polarity: ClockPolarity) -> Result<(), ErrorCode>;
     fn get_clock(&self) -> ClockPolarity;
-    fn set_phase(&self, phase: ClockPhase);
+    fn set_phase(&self, phase: ClockPhase) -> Result<(), ErrorCode>;
```
What if it's an I2C-to-SPI bridge? 

https://www.nxp.com/products/peripherals-and-logic/signal-chain/bridges/ic-bus-to-spi-bridge:SC18IS602B

But I agree with you Brad, I think there shouldn't be errors. This is a pretty tiny edge case, and we don't want to add code size/complexity everywhere for it. It would be fine to have a different SPI interface for an I2C-to-SPI bridge, for which a (software) bridge capsule could then filter out errors if you wanted to use the standard SPIMaster API.

phil-levis commented:
> On some boards `init` performs `configure` which can fail. Should we just ignore that?

They need to be separated. `init` is a one-time operation, while `configure` isn't. We should the code to not have `init` invoke `configure`. Can you point me where?

phil-levis commented:
>     2\. I modified the `sam4l`'s `specify_chip_select` function to report an error if the chip select peripheral is wrong, @phil-levis can you take a look at it as you are the author of the file?

This looks right to me. It probably would be better if the SAM4L didn't use a u8 (this is a very early implementation of this approach), but let's not block on that improvement.

phil-levis commented:
>     3\. The SPI Mux might not be correct for `configure`-like functions. Each time a virtual device asks a configure option, the Mux queues the action and executes it as soon as the real device is idle. Shouldn't these configurations be issues before each data transfer? As it is now, each virtual device can specify a different configuration, but the latest one will be used by all.

Yeah, this is clearly not correct. Each client should have a stored configuration, which is applied when it performs an operation. Calling `configure` merely changes the configuration. This does start to get into the issue of being unable to atomically perform multiple SPI operations (e.g., with a configuration in between); doing that requires explicit split-phase locks ("shared" abstractions in http://sing.stanford.edu/pubs/sosp07-klues.pdf )

alexandruradovici commented:
@phil-levis I gave you access to the repository (https://github.com/WyliodrinEmbeddedIoT/tock), feel free to push any changes.

phil-levis commented:
OK, I think this is ready to go. In fixing the virtualizer I also fixed a long-standing bug in SPI, based on old client configuration approaches. The SPI traits didn't have a `set_client`, and calling `set_client` on virtualized clients did two things: set the callback **and** insert the virtualized client into the list in the MUX. Therefore, if you called `set_client` twice you could corrupt the MUX linked list.

bradjc reviewed (commented):
I personally would rather wait on this until we can add comments to the hil (or at least the functions that changed). SPI was originally written a long time ago and is in a state of neglect.

But I do not want to block this if we are going to get this in for 2.0 testing.

hudson-ayers reviewed (approved):
A few (optional) nits. This looks like strictly an improvement (adds tests, fixes bugs, adds error handling) even if it is not perfect.

phil-levis commented:
> I personally would rather wait on this until we can add comments to the hil (or at least the functions that changed). SPI was originally written a long time ago and is in a state of neglect.
> 
> But I do not want to block this if we are going to get this in for 2.0 testing.

Adding comments, and a TRD, is a good idea. I am happy to put that as a blocking item for a full release (but not testing).

phil-levis commented on boards/imix/src/test/spi_loopback.rs:
```diff
@@ -0,0 +1,133 @@
+    }
+}
+
+#[inline(never)]
```
Yes, it helps with debugging. Trying to sift through the reset handler to figure out your test is the cause of the panic is no fun.

hudson-ayers replied:
makes sense!

phil-levis commented on chips/sam4l/src/spi.rs:
```diff
@@ -528,32 +536,37 @@ impl spi::SpiMaster for SpiHw {
+            }
+            Err(err) => Err(err),
+        }
     }
```
Done.

phil-levis commented on chips/stm32f303xc/src/spi.rs:
```diff
@@ -433,61 +440,72 @@ impl<'a> spi::SpiMaster for Spi<'a> {
+                Ok(self.registers.dr.read(DR::DR) as u8)
+            }
+            Err(err) => Err(err),
+        }
```
Done.

alexandruradovici commented:
Thank you @phil-levis for all the changes. Looks good to me too.

alexandruradovici commented:
as discused in the weekly meeting, for now I added the panic! if init fails in the `SpiMuxComponent`. Ideally this will be replaced with a dedicated panic function.

hudson-ayers commented:
bors r+

@phil-levis if you want an actual release to block on a TRD for SPI, please add that to the Tock 2.0 release issue
